### PR TITLE
coredump: drivers: Add coredump device

### PR DESCRIPTION
**Pull request by mrkhldn**
Add a pseudo device diver with device tree bindings for coredump.
The device tree bindings exposes memory address/size values to be
included in any dump. And the driver exposes an API to add/remove
dump memory regions at runtime.

A test is added that uses the new device and verifies that all
desired memory is included in the core dump when a crash occurs.

Signed-off-by: Mark Holden <mholden@fb.com>

### CODE REVIEW

**Comment by mrkhldn** [comment]
@cfriedt I took an initial pass at a device as suggested here - https://github.com/zephyrproject-rtos/zephyr/pull/41423#issuecomment-1101803926.

Is this what you had in mind?

**Comment by mrkhldn** [comment]
> So say that we had a device tree with a bunch of other arbitrary devices. If we were to add a phandle between a concrete device and the coredump pseudo-device, which would be the parent / child in that relationship? Does it make sense to keep the relationship 1:1?

My thinking is that the coredump pseudo-device would typically be the child device, and a 1:1 relationship makes the most sense. I'm not sure if there's any reason to try and enforce that though.

**Comment by dcpleung** [comment]
IIRC, this kind of pseudo nodes needs to have `zephyr,` as prefix for the `compatible`.

File: MAINTAINERS.yml [hunk 1]
```diff
@@ -544,6 +545,18 @@ Documentation:
     labels:
         - "area: Console"
 
+"Drivers: Coredump Device":
+    status: maintained
+    maintainers:
+        - mrkhldn
+    files:
+        - drivers/coredump_device/
```
Comment (cfriedt) [inline MAINTAINERS.yml:553]: Could probably just shorten this to `drivers/coredump`

**Comment by cfriedt** [comment]
Just one other nitpick - normally, we separate code and tests into separate commits.

Thanks for pulling this together!

**Review by cfriedt — DISMISSED** [review]
minor changes, otherwise, LGTM 👍 
* shorten `coredump_device` to just `coredump`
* split into code commit + test commit

**Review by dcpleung — DISMISSED** [review]

File: tests/drivers/coredump/coredump_api/src/main.c [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+#endif
+
+	return NULL;
+}
+
+ZTEST_SUITE(coredump_tests, NULL, coredump_tests_suite_setup, NULL, NULL, NULL);
+
+ZTEST(coredump_tests, test_register_memory)
+{
+	const struct device *coredump_dev = device_get_binding(COREDUMP_DEVICE_0_NAME);
```
Comment (carlocaione) [inline tests/drivers/coredump/coredump_api/src/main.c:56]: use `DEVICE_DT_GET` instead.

File: drivers/coredump/coredump_impl.c [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+};
+
+#define INIT_REGION(node_id, prop, idx) DT_PROP_BY_IDX(node_id, prop, idx),
+
+#define CREATE_COREDUMP_DEVICE(n)                                                \
+	static struct coredump_data coredump_data_##n;                               \
+	static const struct coredump_config coredump_config##n = {                   \
+		COND_CODE_1(DT_NODE_HAS_PROP(DT_INST(n, DT_DRV_COMPAT), memory_regions), \
+		(                                                                        \
+			.length = DT_PROP_LEN(DT_INST(n, DT_DRV_COMPAT), memory_regions),    \
```
Comment (carlocaione) [inline drivers/coredump/coredump_impl.c:92]: can you avoid using the `COND_CODE_1` and use `DT_PROP_LEN_OR` instead?

File: drivers/coredump/coredump_impl.c [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+	 * Add each memory region specified in device tree to the core dump,
+	 * the memory_regions array should contain two entries per region
+	 * containing the start address and size.
+	 */
+	if ((config->length > 0) && ((config->length % 2) == 0)) {
+		for (int i = 0; i < config->length; i += 2) {
+			uintptr_t start_address = config->memory_regions[i];
+			size_t size = config->memory_regions[i+1];
+
+			coredump_memory_dump(start_address, start_address + size);
```
Comment (carlocaione) [inline drivers/coredump/coredump_impl.c:39]: this is not going to work for MMU-enabled platforms because you are getting the phys from the DT and you are trying to dump that directly. You should at least trying to MMU-map the address before calling the dump function.

**Review by carlocaione — CHANGES_REQUESTED** [review]

File: drivers/coredump/coredump_impl.c (hunk 3 above)
Comment (mrkhldn) [reply drivers/coredump/coredump_impl.c:92]: I could use it for the .length property, but I would still need COND_CODE_1 wrapping the DT_INST_FOREACH_PROP_ELEM when building the memory_regions array because that property is optional

File: drivers/coredump/coredump_impl.c [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+	if ((config->length > 0) && ((config->length % 2) == 0)) {
+		for (int i = 0; i < config->length; i += 2) {
+			uintptr_t start_address = config->memory_regions[i];
+			size_t size = config->memory_regions[i+1];
+
+#if defined(CONFIG_MMU)
+			mm_reg_t virt_addr;
+
+			device_map(&virt_addr, start_address, size, K_MEM_CACHE_NONE);
+			coredump_memory_dump(virt_addr, virt_addr + size);
```
Comment (carlocaione) [inline drivers/coredump/coredump_impl.c:43]: This is wrong. You do not want to map the MMIO region again using `device_map`, this is supposedly done by the driver at init time. You should use `DEVICE_MMIO_GET` instead.

**Review by carlocaione — CHANGES_REQUESTED** [review]

File: drivers/coredump/coredump_impl.c (hunk 4 above)
Comment (cfriedt) [reply drivers/coredump/coredump_impl.c]: I would suggest 2 possible cases for the dt bindings (as enums)
1. simple static address + size (where the coredump subsys just does a memcpy)
2. size + other device with phandle registers a `dump()` callback (would require early init for coredump devices)

File: dts/bindings/coredump/zephyr,coredump.yaml [hunk 6]
```diff
@@ -0,0 +1,16 @@
+# Copyright Meta Platforms, Inc. and its affiliates.
+# SPDX-License-Identifier: Apache-2.0
+
+include: base.yaml
+
+compatible: "zephyr,coredump"
+
+description: Pseudo-device to help capturing desired data into core dumps
+
+properties:
```
Comment (cfriedt) [inline dts/bindings/coredump/zephyr,coredump.yaml:10]: I would suggest 2 possible cases for the dt bindings (as enums)
1. `COREDUMP_TYPE_MEMCPY` (array of memory regions would work)
2. `COREDUMP_TYPE_CALLBACK` (desired size of the coredump memory, and other code may register a callback)

E.g.
```
/ {
	coredump_device0: coredump-device0 {
		compatible = "zephyr,coredump";
		label = "coredump_device_0";
		status = "okay";
    type = COREDUMP_TYPE_CALLBACK;
    /* coredump device could statically allocate array of sufficient size (42) 
     */
    memory-regions = <0x00000000 42>;
	};

	other_device1: other-device1 {
		compatible = "other,device";
		label = "other_device_1";
		status = "okay";
		/* this device (or any arbitrary code that opens the 	coredump_device0 handle)
		 * could register a callback for application-specific coredump behaviour */
    coredump-device = &coredump_device0;
	};
};
```

File: drivers/coredump/coredump_impl.c (hunk 5 above)
Comment (mrkhldn) [reply drivers/coredump/coredump_impl.c:43]: Yeah, I think I understand better now. Let me take Chris's suggestion and primarily use the dt bindings for "simple" memcpy operations at dump time and provide an optional callback so consumers can do more complex operations themselves (like DEVICE_MMIO_GET), that callback can be invoked at dump time

**Review by cfriedt — APPROVED** [review]
LGTM 👍

**Review by carlocaione — APPROVED** [review]

**Review by dcpleung — APPROVED** [review]
